dss/api/collections.py

```diff
@@ -0,0 +1,191 @@
+
+    tombstone_key = CollectionTombstoneID(uuid, version=None).to_key()
+    if test_object_exists(handle, bucket, tombstone_key):
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+
+    if version is None:
+        # list the collections and find the one that is the most recent.
+        prefix = CollectionFQID(uuid, version=None).to_key_prefix()
+        for matching_key in handle.list(bucket, prefix):
+            matching_key = matching_key[len(prefix):]
+            if version is None or matching_key > version:
+                version = matching_key
+    try:
+        collection_blob = handle.get(bucket, CollectionFQID(uuid, version).to_key())
+    except BlobNotFoundError:
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+    return json.loads(collection_blob)
+
+@dss_handler
+def get(uuid: str, replica: str, version: str = None):
+    authenticated_user_email = request.token_info['email']
+    collection_body = get_impl(uuid=uuid, replica=replica, version=version)
+    if collection_body["owner"] != authenticated_user_email:
+        raise DSSException(requests.codes.forbidden, "forbidden", f"Collection access denied")
+    return collection_body
+
+@dss_handler
+def put(json_request_body: dict, replica: str, uuid: str, version: str):
+    authenticated_user_email = request.token_info["email"]
+    collection_body = dict(json_request_body, owner=authenticated_user_email)
+    uuid = uuid.lower()
+    handle = Config.get_blobstore_handle(Replica[replica])
+    dedup_collection: OrderedDict[int, dict] = OrderedDict()
+    for item in collection_body["contents"]:
+        dedup_collection[hash(tuple(sorted(item.items())))] = item
+    collection_body["contents"] = list(dedup_collection.values())
+    verify_collection(collection_body["contents"], Replica[replica], handle)
+    collection_uuid = uuid if uuid else str(uuid4())
+    if version is not None:
+        # convert it to date-time so we can format exactly as the system requires (with microsecond precision)
+        timestamp = iso8601.parse_date(version)
+    else:
+        timestamp = datetime.datetime.utcnow()
+    collection_version = datetime_to_version_format(timestamp)
+    handle.upload_file_handle(Replica[replica].bucket,
+                              CollectionFQID(collection_uuid, collection_version).to_key(),
+                              io.BytesIO(json.dumps(collection_body).encode("utf-8")))
+    return jsonify(dict(uuid=collection_uuid, version=collection_version)), requests.codes.created
+
+class hashabledict(dict):
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+@dss_handler
+def patch(uuid: str, json_request_body: dict, replica: str, version: str):
+    authenticated_user_email = request.token_info['email']
+
+    uuid = uuid.lower()
+    owner = get_impl(uuid=uuid, replica=replica)["owner"]
```

Comment:
if patch indeed does respect tombstones, you can have race conditions.

Reply:
I'm not sure if I follow. The point of tombstones is to be resilient to race conditions, no? Once an all-versions tombstone is placed, no version of the collection can be retrieved.

(Or are you concerned that the PATCH client will be under the mistaken assumption that since the PATCH request succeeded, its result is available?)

Reply:
I take it back. We're replicating the files blindly, not replicating the operations.